File: clojure/src/git_remote_s3/git.clj
; SPDX-FileCopyrightText: 2023-present Amazon.com, Inc. or its affiliates
;
; SPDX-License-Identifier: Apache-2.0

(ns git-remote-s3.git
  "Thin wrappers around git sub-commands invoked as external processes.
   Mirrors git_remote_s3/git.py."
  (:require [clojure.string :as str])
  (:import [java.io File IOException]
           [java.lang ProcessBuilder ProcessBuilder$Redirect]))

;; ---------------------------------------------------------------------------
;; Error type
;; ---------------------------------------------------------------------------

(defn git-error
  "Creates an ex-info map tagged as a git error."
  [message]
  (ex-info message {:type :git-error}))

(defn git-error?
  "Returns true if `e` is a git error thrown by this namespace."
  [e]
  (= :git-error (:type (ex-data e))))

;; ---------------------------------------------------------------------------
;; Private helpers
;; ---------------------------------------------------------------------------

(defn- run-process
  "Runs `args` as an external command.

  Options:
    :redirect-stdout - one of :pipe (default), :inherit, :stderr
    :redirect-stderr - one of :pipe (default), :inherit, :discard
    :check           - when true (default), throws GitError on non-zero exit

  Returns a map {:exit <int> :out <String> :err <String>}."
  [args & {:keys [redirect-stdout redirect-stderr check]
           :or   {redirect-stdout :pipe
                  redirect-stderr :pipe
                  check           true}}]
  (let [pb (ProcessBuilder. ^java.util.List (map str args))]
    ;; stdout
    (case redirect-stdout
      :pipe    (.redirectOutput pb ProcessBuilder$Redirect/PIPE)
      :inherit (.redirectOutput pb ProcessBuilder$Redirect/INHERIT)
      :stderr  (.redirectOutput pb ProcessBuilder$Redirect/INHERIT))
    ;; stderr
    (case redirect-stderr
      :pipe    (.redirectError pb ProcessBuilder$Redirect/PIPE)
      :inherit (.redirectError pb ProcessBuilder$Redirect/INHERIT)
      :discard (.redirectError pb ProcessBuilder$Redirect/DISCARD))
    (let [proc (.start pb)
          out  (String. (.readAllBytes (.getInputStream proc)) "UTF-8")
          err  (String. (.readAllBytes (.getErrorStream proc)) "UTF-8")
          exit (.waitFor proc)]
      (when (and check (not (zero? exit)))
        (throw (git-error (if (str/blank? err) out err))))
      {:exit exit :out out :err err})))

;; ---------------------------------------------------------------------------
;; Public API
;; ---------------------------------------------------------------------------

(defn archive
  "Archives the content of the repo into <folder>/repo.zip using `git archive`.

  Returns the path to the archive file."
  [& {:keys [folder ref]}]
  (let [file-path (str folder "/repo.zip")]
    (run-process ["git" "archive" "--format" "zip" "--output" file-path ref])
    file-path))

(defn bundle
  "Bundles a ref into <folder>/<sha>.bundle using `git bundle create`.

  Returns the path to the bundle file."
  [& {:keys [folder sha ref]}]
  (let [file-path (str folder "/" sha ".bundle")]
    (run-process ["git" "bundle" "create" file-path ref])
    file-path))

(defn unbundle
  "Unbundles <folder>/<sha>.bundle for `ref` using `git bundle unbundle`.

  Stdout of the sub-process is forwarded to the JVM's stderr (matching
  the Python `stdout=sys.stderr` behaviour)."
  [& {:keys [folder sha ref]}]
  ;; git bundle unbundle writes progress to stdout; we redirect that to
  ;; the process's inherited stderr so it reaches the user via System.err.
  (let [pb (ProcessBuilder. ^java.util.List
                            ["git" "bundle" "unbundle"
                             (str folder "/" sha ".bundle") ref])]
    ;; Both stdout and stderr go to the parent's stderr
    (.redirectOutput pb ProcessBuilder$Redirect/INHERIT)
    (.redirectError  pb ProcessBuilder$Redirect/INHERIT)
    (let [proc (.start pb)
          exit (.waitFor proc)]
      (when-not (zero? exit)
        (throw (git-error (str "git bundle unbundle failed with exit " exit)))))))

(defn rev-parse
  "Returns the full SHA for `ref`, or throws a git-error if not found."
  [ref]
  (let [{:keys [exit out]} (run-process ["git" "rev-parse" ref]
                                        :check false)]
    (if (zero? exit)
      (str/trim out)
      (throw (git-error (str "fatal: " ref " not found"))))))

(defn is-ancestor
  "Returns true when `ancestor` is a git ancestor of `descendant`."
  [ancestor descendant]
  (let [{:keys [exit]} (run-process
                        ["git" "merge-base" "--is-ancestor" ancestor descendant]
                        :check           false
                        :redirect-stdout :discard
                        :redirect-stderr :discard)]
    (zero? exit)))

(defn get-remote-url
  "Returns the URL for `remote`, or throws a git-error if not found."
  [remote]
  (let [{:keys [exit out]} (run-process ["git" "remote" "get-url" remote]
                                        :check false)]
    (if (zero? exit)
      (str/trim out)
      (throw (git-error (str "fatal: " remote " not found"))))))

(defn validate-ref-name
  "Returns true when `name` is a valid git ref name.

  Mirrors the Python implementation which follows:
  https://github.com/git/git/blob/406f326d/refs.c#L170"
  [name]
  (nil? (re-find
         #"(^\.)|(\.\.)|([:\?\[\\\^\~\s\*\]])|(\.lock$)|(/$)|(@\{)|([\x00-\x1f])"
         name)))

(defn get-last-commit-message
  "Returns the abbreviated hash and subject of the last commit."
  []
  (let [{:keys [exit out]} (run-process ["git" "log" "-1" "--pretty=%h %s"]
                                        :check false)]
    (if (zero? exit)
      (str/trim out)
      (throw (git-error "fatal: an error has occurred")))))

File: elixir/lib/git_remote_s3/git.ex
# SPDX-FileCopyrightText: 2023-present Amazon.com, Inc. or its affiliates
#
# SPDX-License-Identifier: Apache-2.0

defmodule GitRemoteS3.GitError do
  @moduledoc "Raised when a git subprocess returns a non-zero exit code."
  defexception [:message]
end

defmodule GitRemoteS3.Git do
  @moduledoc """
  Thin wrappers around git subprocess commands.

  Mirrors `git_remote_s3/git.py`.  All functions raise `GitRemoteS3.GitError`
  on failure.
  """

  alias GitRemoteS3.GitError

  # ---------------------------------------------------------------------------
  # archive
  # ---------------------------------------------------------------------------

  @doc """
  Archive the working tree at `ref` into `<folder>/repo.zip`.

  Returns the path to the archive file, or raises `GitError`.
  """
  @spec archive(folder: String.t(), ref: String.t()) :: String.t()
  def archive(folder: folder, ref: ref) do
    file_path = Path.join(folder, "repo.zip")

    case System.cmd("git", ["archive", "--format", "zip", "--output", file_path, ref],
           stderr_to_stdout: false
         ) do
      {_out, 0} ->
        file_path

      {err, _code} ->
        raise GitError, message: err
    end
  end

  # ---------------------------------------------------------------------------
  # bundle
  # ---------------------------------------------------------------------------

  @doc """
  Bundle `ref` into `<folder>/<sha>.bundle`.

  Returns the path to the bundle file, or raises `GitError`.
  """
  @spec bundle(folder: String.t(), sha: String.t(), ref: String.t()) :: String.t()
  def bundle(folder: folder, sha: sha, ref: ref) do
    file_path = Path.join(folder, "#{sha}.bundle")

    case System.cmd("git", ["bundle", "create", file_path, ref], stderr_to_stdout: false) do
      {_out, 0} ->
        file_path

      {err, _code} ->
        raise GitError, message: err
    end
  end

  # ---------------------------------------------------------------------------
  # unbundle
  # ---------------------------------------------------------------------------

  @doc """
  Unbundle `<folder>/<sha>.bundle` and check out `ref`.

  Output is forwarded to stderr (matching the Python implementation's
  `stdout=sys.stderr`).
  """
  @spec unbundle(folder: String.t(), sha: String.t(), ref: String.t()) :: :ok
  def unbundle(folder: folder, sha: sha, ref: ref) do
    bundle_path = Path.join(folder, "#{sha}.bundle")

    # System.cmd captures stdout; we write it to :stderr ourselves to mirror
    # the Python `stdout=sys.stderr` behaviour.
    case System.cmd("git", ["bundle", "unbundle", bundle_path, ref], stderr_to_stdout: false) do
      {output, 0} ->
        IO.write(:stderr, output)
        :ok

      {err, code} ->
        raise GitError, message: "git bundle unbundle exited with #{code}: #{err}"
    end
  end

  # ---------------------------------------------------------------------------
  # rev_parse
  # ---------------------------------------------------------------------------

  @doc """
  Return the SHA for `ref`, or raise `GitError` if not found.
  """
  @spec rev_parse(String.t()) :: String.t()
  def rev_parse(ref) do
    case System.cmd("git", ["rev-parse", ref], stderr_to_stdout: false) do
      {sha, 0} ->
        String.trim(sha)

      {_err, _code} ->
        raise GitError, message: "fatal: #{ref} not found"
    end
  end

  # ---------------------------------------------------------------------------
  # is_ancestor
  # ---------------------------------------------------------------------------

  @doc """
  Return `true` if `ancestor` is an ancestor of `descendant`.
  """
  @spec is_ancestor(String.t(), String.t()) :: boolean()
  def is_ancestor(ancestor, descendant) do
    {_out, code} =
      System.cmd("git", ["merge-base", "--is-ancestor", ancestor, descendant],
        stderr_to_stdout: true
      )

    code == 0
  end

  # ---------------------------------------------------------------------------
  # get_remote_url
  # ---------------------------------------------------------------------------

  @doc """
  Return the URL of `remote`, or raise `GitError` if not found.
  """
  @spec get_remote_url(String.t()) :: String.t()
  def get_remote_url(remote) do
    case System.cmd("git", ["remote", "get-url", remote], stderr_to_stdout: false) do
      {url, 0} ->
        String.trim(url)

      {_err, _code} ->
        raise GitError, message: "fatal: #{remote} not found"
    end
  end

  # ---------------------------------------------------------------------------
  # validate_ref_name
  # ---------------------------------------------------------------------------

  # Mirrors https://github.com/git/git/blob/406f326/refs.c#L170
  @invalid_ref_regex ~r/(\A\.)|(\.\.)|([:\?\[\\\^\~\s\*\]])|(\.lock\z)|(\/\z)|(@\{)|([\x00-\x1f])/

  @doc """
  Return `true` if `name` is a valid git ref name.
  """
  @spec validate_ref_name(String.t()) :: boolean()
  def validate_ref_name(name) do
    not Regex.match?(@invalid_ref_regex, name)
  end

  # ---------------------------------------------------------------------------
  # get_last_commit_message
  # ---------------------------------------------------------------------------

  @doc """
  Return the short log line for the most recent commit, or raise `GitError`.
  """
  @spec get_last_commit_message() :: String.t()
  def get_last_commit_message do
    case System.cmd("git", ["log", "-1", "--pretty=%h %s"], stderr_to_stdout: false) do
      {msg, 0} ->
        String.trim(msg)

      {_err, _code} ->
        raise GitError, message: "fatal: an error as occurred"
    end
  end
end
